### PR TITLE
V5 STJ Optional<T> `JsonIgnoreAttribute` Analyzer

### PR DESCRIPTION
**Pull request by OoLunar**
# Summary
Adds the `DSharpPlus.Analyzers` and `DSharpPlus.Test` projects.

# Details
Inside the `DSharpPlus.Analyzers` project is a singular analyzer with an associated code fix. `DSP0001` is intended to be used until STJ adds support for conditionally skipping properties as claimed in https://github.com/dotnet/runtime/issues/36275. The analyzer will point out `Optional<T>` properties without `[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]` and emit a warning (or error, as defined in our `.editorconfig`). The code fix will add the required `using System.Text.Json.Serialization;` directive on an "as needed" basis.
Since the DSP0001 analyzer has a code fix available, it is compatible with `dotnet format`. Github Actions can utilize `dotnet format` on new PRs or new commits. This means that PR's adding new optional properties won't be required to explicitly add `[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]`, as `dotnet format` will do it for you. In the future, `dotnet format` is planned to be run by Github Actions on newly opened PRs or on new commits.

Inside `DSharpPlus.Test`, each project will be separated by folders/namespaces where each element of the project can be tested using the test SDK (compatible with `dotnet test`). Because `DSharpPlus.Test` can utilize `dotnet test`, we can integrate it with Github Actions for further use down the line.
As it currently stands, the `DSharpPlus.Test` project only has one file with five analyzer tests inside. Testing the analyzer should always be done to prevent false positives or potential `dotnet format` breakages.

# Notes
Conflicts with #1260. #1260 adds a [`ReflectJsonConverterFactory.cs`](https://github.com/DSharpPlus/DSharpPlus/pull/1260/files#diff-4fecbb95eb1095ab61910653a0cbc32b8ae3847dfc4667ddd564ac32f2911787) class which utilizes reflection and IL to conditionally skip properties.

The main difference between #1260 and this PR is readability vs. performance.

### CODE REVIEW

**Comment by OoLunar** [comment]
```
[lunar@J-Laptop 10:18:21pm ~/Code/DSharpPlus/rewrite_5.0]$ dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  DSharpPlus.Analyzers -> /home/lunar/Code/DSharpPlus/rewrite_5.0/DSharpPlus.Analyzers/bin/Debug/net6.0/DSharpPlus.Analyzers.dll
  DSharpPlus.Test -> /home/lunar/Code/DSharpPlus/rewrite_5.0/DSharpPlus.Test/bin/Debug/net6.0/DSharpPlus.Test.dll
Test run for /home/lunar/Code/DSharpPlus/rewrite_5.0/DSharpPlus.Test/bin/Debug/net6.0/DSharpPlus.Test.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.1.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:     5, Skipped:     0, Total:     5, Duration: 4 s - /home/lunar/Code/DSharpPlus/rewrite_5.0/DSharpPlus.Test/bin/Debug/net6.0/DSharpPlus.Test.dll (net6.0)
```
The PR is in a state ready for merging. Whether we accept https://github.com/DSharpPlus/DSharpPlus/pull/1260 or this PR is a matter to be discussed on the Discord. It is recommend to run benchmarks before starting any discussion.

**Comment by OoLunar** [comment]
Requesting a review from @DPlayer234

**Comment by OoLunar** [comment]
Found a code replacement bug when attempting to replace an already existing `JsonIgnoreAttribute(Condition = JsonIgnoreCondition.WhenWritingAnythingExceptDefault)`. A new test and fix will be added, along with redundant code comments being removed.

File: DSharpPlus.Analyzers/Core/DSP0001/OptionalAddJsonIgnoreCodeFixProvider.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                title: CodeFixTitle,
+                createChangedDocument: c => AddJsonIgnoreAttributeAsync(context.Document, declaration, c),
+                equivalenceKey: CodeFixTitle), diagnostic);
+        }
+
+        private static async Task<Document> AddJsonIgnoreAttributeAsync(Document document, PropertyDeclarationSyntax propertyDeclarationSyntax, CancellationToken cancellationToken)
+        {
+            SyntaxNode root = await document.GetSyntaxRootAsync(cancellationToken);
+            Document newDocument = AddAttributes(document, root, propertyDeclarationSyntax);
+            return newDocument.WithSyntaxRoot(await AddUsings((await newDocument.GetSyntaxRootAsync(cancellationToken)).SyntaxTree.GetCompilationUnitRoot(cancellationToken)).GetRootAsync(cancellationToken));
```
Comment (DPlayer234) [inline DSharpPlus.Analyzers/Core/DSP0001/OptionalAddJsonIgnoreCodeFixProvider.cs:62]: The 2 awaits in this expression make it fairly hard to read.

File: DSharpPlus.Analyzers/Core/DSP0001/OptionalAddJsonIgnoreCodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            //    jsonIgnoreAttribute = jsonIgnoreAttribute.WithTriviaFrom(firstAttribute);
+            //    return document.WithSyntaxRoot(root.ReplaceNode(propertyDeclarationSyntax, propertyDeclarationSyntax.WithAttributeLists(propertyDeclarationSyntax.AttributeLists.Replace(firstAttribute, firstAttribute.WithoutTrivia()).Insert(0, SyntaxFactory.AttributeList(SyntaxFactory.SingletonSeparatedList(jsonIgnoreAttribute))))));
+            //}
+            //else
+            //{
+            //}
+#pragma warning disable IDE0046 // Convert to conditional expression
+            if (propertyDeclarationSyntax.AttributeLists.Any())
+            {
+                return document.WithSyntaxRoot(root.ReplaceNode(propertyDeclarationSyntax, propertyDeclarationSyntax.WithAttributeLists(propertyDeclarationSyntax.AttributeLists.Add(SyntaxFactory.AttributeList(SyntaxFactory.SingletonSeparatedList(jsonIgnoreAttribute))))));
```
Comment (DPlayer234) [inline DSharpPlus.Analyzers/Core/DSP0001/OptionalAddJsonIgnoreCodeFixProvider.cs:87]: The expression `SyntaxFactory.AttributeList(SyntaxFactory.SingletonSeparatedList(jsonIgnoreAttribute))` seems to be present in each of these branches. That should probably be factored out. Adding some appropriate line breaks would also help with readability.

**Review by DPlayer234 — APPROVED** [review]
Don't see anything wrong overall. Just some nitpicking about style.

File: DSharpPlus.Analyzers/Core/DSP0001/OptionalAddJsonIgnoreCodeFixProvider.cs (hunk 1 above)
Comment (OoLunar) [reply DSharpPlus.Analyzers/Core/DSP0001/OptionalAddJsonIgnoreCodeFixProvider.cs:62]: Fixed in next commit

File: DSharpPlus.Analyzers/Core/DSP0001/OptionalAddJsonIgnoreCodeFixProvider.cs (hunk 2 above)
Comment (OoLunar) [reply DSharpPlus.Analyzers/Core/DSP0001/OptionalAddJsonIgnoreCodeFixProvider.cs:87]: Looking at the current commit, how would you think I should go about this?

Comment (OoLunar) [reply DSharpPlus.Analyzers/Core/DSP0001/OptionalAddJsonIgnoreCodeFixProvider.cs:87]: Will do! Thanks for your input

**Pull request closed without merging**